spec.html

```diff
@@ -14189,6 +14206,53 @@ <h1>Runtime Semantics: Evaluation</h1>
         </emu-alg>
+
+        <emu-grammar>ImportMeta : `import` `.` `meta`</emu-grammar>
+        <emu-alg>
+          1. Let _module_ be GetActiveScriptOrModule().
+          1. Assert: _module_ is a Source Text Module Record.
+          1. Let _importMeta_ be _module_.[[ImportMeta]].
+          1. If _importMeta_ is ~empty~, then
+            1. Set _importMeta_ to ! OrdinaryObjectCreate(*null*).
+            1. Let _importMetaValues_ be ! HostGetImportMetaProperties(_module_).
+            1. For each Record { [[Key]], [[Value]] } _p_ that is an element of _importMetaValues_,
+              1. Perform ! CreateDataPropertyOrThrow(_importMeta_, _p_.[[Key]], _p_.[[Value]]).
+            1. Perform ! HostFinalizeImportMeta(_importMeta_, _module_).
+            1. Set _module_.[[ImportMeta]] to _importMeta_.
+            1. Return _importMeta_.
+          1. Else,
+            1. Assert: Type(_importMeta_) is Object.
+            1. Return _importMeta_.
+        </emu-alg>
+
+        <emu-clause id="sec-hostgetimportmetapropeties" aoid="HostGetImportMetaProperties">
+          <h1>Runtime Semantics: HostGetImportMetaProperties ( _moduleRecord_ )</h1>
+          <p>HostGetImportMetaProperties is an implementation-defined abstract operation that allows hosts to provide property keys and values for the object returned from `import.meta`.</p>
+
+          <p>The implementation of HostGetImportMetaProperties must conform to the following requirements:</p>
+          <ul>
+            <li>It must return a List, whose values are all Records with two fields, [[Key]] and [[Value]].</li>
+            <li>Each such Record's [[Key]] field must be a property key, i.e., IsPropertyKey must return *true* when applied to it.</li>
+            <li>Each such Record's [[Value]] field must be an ECMAScript value.</li>
+            <li>It must always complete normally (i.e., not return an abrupt completion).</li>
+          </ul>
+
+          <p>The default implementation of HostGetImportMetaProperties is to return a new empty List.</p>
+        </emu-clause>
+
+        <emu-clause id="sec-hostfinalizeimportmeta" aoid="HostFinalizeImportMeta">
+          <h1>Runtime Semantics: HostFinalizeImportMeta ( _importMeta_, _moduleRecord_ )</h1>
+          <p>HostFinalizeImportMeta is an implementation-defined abstract operation that allows hosts to perform any extraordinary operations to prepare the object returned from `import.meta`.</p>
+
+          <p>Most hosts will be able to simply define HostGetImportMetaProperties, and leave HostFinalizeImportMeta with its default behavior. However, HostFinalizeImportMeta provides an "escape hatch" for hosts which need to directly manipulate the object before it is exposed to ECMAScript code.</p>
+
+          <p>The implementation of HostFinalizeImportMeta must conform to the following requirements:</p>
+          <ul>
+            <li>It must always complete normally (i.e., not return an abrupt completion).</li>
```

Comment:
We should probably say more about what it can do to the _importMeta_ object that's passed in? It shouldn't, e.g., transmute it into an exotic object by mucking with the internal methods.

Reply:
Sounds like that might need to be a more general forbidden extension then?

Reply:
Yeah, the specific case of transmutation into an exotic sounds like a good general forbidden extension to have. Did we discuss the general invariants of this host hook?

Reply:
I think as long as this hook doesn't make the object exotic, anything is fair game, including changing [[Prototype]].

Reply:
https://github.com/tc39/ecma262/pull/1892#discussion_r389292774

Reply:
> I think as long as this hook doesn't make the object exotic, anything is fair game, including changing [[Prototype]].

That seems reasonable. Is this captured in the notes somewhere?

Reply:
@syg i couldn't find anything as explicit as my sentence above, but between issues and notes and light discussion with domenic that seems to be the understanding.